File: ComponentImpl/src/main/java/com/xiaojinzi/component/impl/fragment/FragmentCenter.java
package com.xiaojinzi.component.impl.fragment;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.xiaojinzi.component.Component;
import com.xiaojinzi.component.ComponentUtil;
import com.xiaojinzi.component.anno.support.CheckClassNameAnno;
import com.xiaojinzi.component.fragment.IComponentCenterFragment;
import com.xiaojinzi.component.fragment.IComponentHostFragment;
import com.xiaojinzi.component.support.ASMUtil;
import com.xiaojinzi.component.support.Utils;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * 模块 Fragment 注册和卸载的总管
 *
 * @author xiaojinzi
 */
@CheckClassNameAnno
public class FragmentCenter implements IComponentCenterFragment {

    private Map<String, IComponentHostFragment> moduleFragmentMap = new HashMap<>();

    private FragmentCenter() {
    }

    private static volatile FragmentCenter instance;

    public static FragmentCenter getInstance() {
        if (instance == null) {
            synchronized (FragmentCenter.class) {
                if (instance == null) {
                    instance = new FragmentCenter();
                }
            }
        }
        return instance;
    }

    @Override
    public void register(@NonNull IComponentHostFragment hostFragment) {
        Utils.checkNullPointer(hostFragment);
        if (!moduleFragmentMap.containsKey(hostFragment.getHost())) {
            moduleFragmentMap.put(hostFragment.getHost(), hostFragment);
            hostFragment.onCreate(Component.getApplication());
        }
    }

    @Override
    public void register(@NonNull String host) {
        Utils.checkStringNullPointer(host, "host");
        if (!moduleFragmentMap.containsKey(host)) {
            IComponentHostFragment moduleService = findModuleService(host);
            if (moduleService != null) {
                register(moduleService);
            }
        }
    }

    @Override
    public void unregister(@NonNull IComponentHostFragment hostFragment) {
        Utils.checkNullPointer(hostFragment);
        moduleFragmentMap.remove(hostFragment.getHost());
        hostFragment.onDestroy();
    }

    @Override
    public void unregister(@NonNull String host) {
        Utils.checkStringNullPointer(host, "host");
        IComponentHostFragment moduleService = moduleFragmentMap.get(host);
        if (moduleService != null) {
            unregister(moduleService);
        }
    }

    @Nullable
    public IComponentHostFragment findModuleService(String host) {
        try {
            if (Component.getConfig().isOptimizeInit()) {
                return ASMUtil.findModuleFragmentAsmImpl(
                        ComponentUtil.transformHostForClass(host)
                );
            } else {
                Class<? extends IComponentHostFragment> clazz = null;
                String className = ComponentUtil.genHostFragmentClassName(host);
                clazz = (Class<? extends IComponentHostFragment>) Class.forName(className);
                return clazz.newInstance();
            }
        } catch (Exception ignore) {
            // ignore
        }
        return null;
    }

    public void check() {
        Set<String> set = new HashSet<>();
        for (Map.Entry<String, IComponentHostFragment> entry : moduleFragmentMap.entrySet()) {
            IComponentHostFragment childRouter = entry.getValue();
            if (childRouter == null || childRouter.getFragmentNameSet() == null) {
                continue;
            }
            Set<String> childRouterSet = childRouter.getFragmentNameSet();
            for (String key : childRouterSet) {
                if (set.contains(key)) {
                    throw new IllegalStateException("the name of Fragment is exist：'" + key + "'");
                }
                set.add(key);
            }
        }
    }

}

File: ComponentImpl/src/main/java/com/xiaojinzi/component/impl/fragment/FragmentCenter.kt
package com.xiaojinzi.component.impl.fragment

import com.xiaojinzi.component.Component.getApplication
import com.xiaojinzi.component.Component.requiredConfig
import com.xiaojinzi.component.ComponentUtil
import com.xiaojinzi.component.anno.support.CheckClassNameAnno
import com.xiaojinzi.component.fragment.IComponentCenterFragment
import com.xiaojinzi.component.fragment.IComponentHostFragment
import com.xiaojinzi.component.support.ASMUtil
import com.xiaojinzi.component.support.Utils
import java.util.*

/**
 * 模块 Fragment 注册和卸载的总管
 *
 * @author xiaojinzi
 */
@CheckClassNameAnno
object FragmentCenter: IComponentCenterFragment {

    private val moduleFragmentMap: MutableMap<String, IComponentHostFragment> = HashMap()

    override fun register(hostFragment: IComponentHostFragment) {
        Utils.checkNullPointer(hostFragment)
        if (!moduleFragmentMap.containsKey(hostFragment.host)) {
            moduleFragmentMap[hostFragment.host] = hostFragment
            hostFragment.onCreate(getApplication())
        }
    }

    override fun register(host: String) {
        Utils.checkStringNullPointer(host, "host")
        if (!moduleFragmentMap.containsKey(host)) {
            val moduleService = findModuleService(host)
            moduleService?.let { register(it) }
        }
    }

    override fun unregister(hostFragment: IComponentHostFragment) {
        Utils.checkNullPointer(hostFragment)
        moduleFragmentMap.remove(hostFragment.host)
        hostFragment.onDestroy()
    }

    override fun unregister(host: String) {
        Utils.checkStringNullPointer(host, "host")
        val moduleService = moduleFragmentMap[host]
        moduleService?.let { unregister(it) }
    }

    fun findModuleService(host: String?): IComponentHostFragment? {
        try {
            return if (requiredConfig().isOptimizeInit) {
                ASMUtil.findModuleFragmentAsmImpl(
                        ComponentUtil.transformHostForClass(host)
                )
            } else {
                var clazz: Class<out IComponentHostFragment?>? = null
                val className = ComponentUtil.genHostFragmentClassName(host)
                clazz = Class.forName(className) as Class<out IComponentHostFragment>
                clazz.newInstance()
            }
        } catch (ignore: Exception) {
            // ignore
        }
        return null
    }

    fun check() {
        val set: MutableSet<String> = HashSet()
        for ((_, childRouter) in moduleFragmentMap) {
            if (childRouter?.fragmentNameSet == null) {
                continue
            }
            val childRouterSet = childRouter.fragmentNameSet
            for (key in childRouterSet) {
                check(!set.contains(key)) { "the name of Fragment is exist：'$key'" }
                set.add(key)
            }
        }
    }

}